clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -1017,34 +1019,61 @@ private void maybeCloseRecordStream() {
             }
         }
 
-        private Record nextFetchedRecord() {
-            if (hasExceptionInLastFetch) {
-                if (lastRecord == null) {
-                    maybeEnsureValid(currentBatch);
+        /**
+         * Because after iterator.hasNext() throws exception, the state of the iterator will be FAILED.
+         * So the exception is not reproducible. Here we cache the first exception thrown from iterator.hasNext() and
+         * use that exception as suppressed exception when the iterator throws exception again.
+         */
+        private boolean wrappedHasNext(Iterator iter) {
+            try {
+                return iter.hasNext();
+            } catch (KafkaException ke) {
+                cachedIteratorException = ke;
+                throw ke;
+            } catch (IllegalStateException ise) {
+                if (cachedIteratorException != null) {
+                    ise.addSuppressed(cachedIteratorException);
+                    throw new KafkaException(ise);
                 } else {
-                    maybeEnsureValid(lastRecord);
-                    return lastRecord;
+                    throw ise;
                 }
             }
+        }
 
+        private Record nextFetchedRecord() {
```

Comment:
I think it would be clearer if we handled exceptions from the iterator and from the crc check consistently. For the iterator case, we cache an exception; for the crc check, we just re-execute `ensureValid`. Perhaps we could cache the exception in both cases?